src/Semantics/Interp.dfy

```diff
@@ -231,14 +242,34 @@ module Bootstrap.Semantics.Interp {
     match e {
       case Var(v) =>
         LiftPureResult(ctx, InterpVar(v, ctx, env))
+
       case Abs(vars, body) =>
         var cv: V.T := V.Closure(ctx.locals, vars, body);
         assert WellFormedValue(cv); // TODO: prove
         Success(Return(cv, ctx))
+
       case Literal(lit) =>
         Success(Return(InterpLiteral(lit), ctx))
+
       case Apply(Lazy(op), args: seq<Expr>) =>
-        InterpLazy(e, env, ctx)
+        // This is ``InterpLazy`` inlined.
+        // See the comments for the ``Block`` case.
+        reveal SupportsInterp();
+        Predicates.Deep.AllImpliesChildren(e, SupportsInterp1);
+        var op, e0, e1 := e.aop.lOp, e.args[0], e.args[1];
+        var Return(v0, ctx0) :- InterpExpr(e0, env, ctx);
+        :- NeedType(e0, v0, Type.Bool);
+        match (op, v0) {
+          case (And, Bool(false)) => Success(Return(V.Bool(false), ctx0))
+          case (Or,  Bool(true))  => Success(Return(V.Bool(true), ctx0))
+          case (Imp, Bool(false)) => Success(Return(V.Bool(true), ctx0))
+          case (_,   Bool(b)) =>
+            assert op in {Exprs.And, Exprs.Or, Exprs.Imp};
+            var Return(v1, ctx1) :- InterpExpr(e1, env, ctx0);
+            :- NeedType(e1, v1, Type.Bool);
+            Success(Return(v1, ctx1))
+        }
```

Comment:
I would like to undertand why this goes wrong before we make the interpreter *more* monolithic; can we maybe add fuel in the right places?

Reply:
In particular, every lemma about this function is going to have a lot more reasoning to do, since its body is so large.  I don't think its sustainable to inline everything.

Reply:
Let's discuss that.
The main issue is that we have to *remember* that in this case there are several levels of recursive calls, otherwise we often have proof obligations which fail and need to debug what is happening (which is expensive). Of course, once the issue is identified (it is a problem of fuel) there are several solutions: we can add the proper recursive calls in the proof, and we can also increase the fuel. This doesn't change the core issue.
About increasing the fuel (generally speaking): I tend not to do that, because when I do it in F* it quickly makes the context explode. Maybe the encoding in Dafny is different, but I prefer to be cautious (we can discuss this).

Reply:
Let's do this later.  We could try increasing the fuel and see if that allows us to make things less monolithic.

Reply:
Let's discuss that with @RustanLeino